tests/test_threadpoolctl.py

```diff
@@ -24,105 +22,124 @@ def effective_num_threads(nthreads, max_threads):
+@pytest.mark.parametrize("prefix", _ALL_PREFIXES)
+@pytest.mark.parametrize("limit", [1, 3])
+def test_threadpool_limits_by_prefix(prefix, limit):
+    # Check that the maximum number of threads can be set by prefix
+    original_infos = _threadpool_info()
+
+    modules_matching_prefix = original_infos.get_modules("prefix", prefix)
+    if not modules_matching_prefix:
+        pytest.skip("Requires {} runtime".format(prefix))
+
+    with threadpool_limits(limits={prefix: limit}):
+        for module in modules_matching_prefix:
             if is_old_openblas(module):
                 continue
-            if module["user_api"] in user_apis:
-                assert module["num_threads"] == 1
+            # threadpool_limits only sets an upper bound on the number of
+            # threads.
+            assert 0 < module.get_num_threads() <= limit
+    assert _threadpool_info() == original_infos
+
 
-    with threadpool_limits(limits=3, user_api=user_api):
-        for module in threadpool_info():
+@pytest.mark.parametrize("user_api", (None, "blas", "openmp"))
+@pytest.mark.parametrize("limit", [1, 3])
+def test_set_threadpool_limits_by_api(user_api, limit):
+    # Check that the maximum number of threads can be set by user_api
+    original_infos = _threadpool_info()
+
+    modules_matching_api = original_infos.get_modules("user_api", user_api)
+    if not modules_matching_api:
+        user_apis = _ALL_USER_APIS if user_api is None else [user_api]
+        pytest.skip("Requires a library which api is in {}".format(user_apis))
+
+    with threadpool_limits(limits=limit, user_api=user_api):
+        for module in modules_matching_api:
             if is_old_openblas(module):
                 continue
-            if module["user_api"] in user_apis:
-                # threadpool_limits only sets an upper bound on the number of
-                # threads.
-                assert module["num_threads"] <= 3
+            # threadpool_limits only sets an upper bound on the number of
+            # threads.
+            assert 0 < module.get_num_threads() <= limit
 
-    assert threadpool_info() == original_infos
+    assert _threadpool_info() == original_infos
 
 
 def test_threadpool_limits_function_with_side_effect():
     # Check that threadpool_limits can be used as a function with
     # side effects instead of a context manager.
-    original_infos = threadpool_info()
+    original_infos = _threadpool_info()
 
     threadpool_limits(limits=1)
     try:
-        for module in threadpool_info():
+        for module in _threadpool_info():
             if is_old_openblas(module):
                 continue
-            assert module["num_threads"] == 1
+            assert module.num_threads == 1
     finally:
         # Restore the original limits so that this test does not have any
         # side-effect.
         threadpool_limits(limits=original_infos)
 
-    assert threadpool_info() == original_infos
+    assert _threadpool_info() == original_infos
```

Comment:
It depends on the test. For tests that can be simplified significantly by using the private API I think it's ok to use it. Maybe we could add a comment to explain why we test against the private API in those cases.

Reply:
Yes I agree for tests that are simplified a lot with the use of `_threadpool_info`. I only meant for tests that have been modified only to accomodate with the use of `is_old_openblas`.

Reply:
I agree that some tests are not simplified with the new object api. I chose to use the private api in all existing tests for consistency because otherwise I think it might be confusing to have some tests using the public and some tests using the private api with no obvious reason to choose one over the other. Of course, in addition, I made new tests to check consistency between the public and private api.